### PR TITLE
Feature/defenders fixes

### PR DESCRIPTION
**Pull request by rolfvdhulst**
Fixes multiple small things/bugs. Branched from #277 so merge that first.

### CODE REVIEW

**Comment by rolfvdhulst** [comment]
Is ready for review, but there's still problems with the pass implementation. I need @RobZelluf 's help a bit.

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp [hunk 1]
```diff
@@ -49,7 +49,7 @@ double DefencePositionCoach::getOrientation(const Line &line) {
 //computes a line segment on which the entirety of openGoalSegment is blocked as seen from point with robots with radius collissionRadius
 std::shared_ptr<Line> DefencePositionCoach::getBlockLineSegment(
         const Line &openGoalSegment, const Vector2 &point, double collisionRadius, double margin) {
-    if (margin == - 1.0) { margin = collisionRadius; }
+    if (margin == - 1.0) { margin = 0.15; }
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: Magic

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp [hunk 2]
```diff
@@ -127,6 +127,7 @@ std::vector<DefencePositionCoach::DefenderBot> DefencePositionCoach::decidePosit
     // we start by creating a simulated world in which we place our robots according to the most needed positions 1 by 1
     world::WorldData simulatedWorld = world::world->getWorld();
     simulatedWorld.us.clear();
+    simulatedWorld = getTheirAttackers(simulatedWorld);    // we select only the relevant robots
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: this method is way too large

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp [hunk 3]
```diff
@@ -322,6 +327,65 @@ Vector2 DefencePositionCoach::findPositionForBlockBall(const Line &blockLine) {
     }
     return getPosOnLine(blockLine, 0.3);
 }
+
+world::WorldData DefencePositionCoach::getTheirAttackers(const world::WorldData &world) {
+    std::vector<world::Robot> theirAttackers;
+    for (const world::Robot &robot :world.them) {
+        // we remove any attackers that are outside of the field or in our defence area
+        if (world::field->pointIsInDefenceArea(robot.pos, true, 0.04)) {
+            continue;
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: continues can give undefined behaviour. it is better not to use them. You could better take the two boolean statements and push attackers into your vector when they are both false.  (also saves 4 lines)

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp [hunk 4]
```diff
@@ -322,6 +327,65 @@ Vector2 DefencePositionCoach::findPositionForBlockBall(const Line &blockLine) {
...
+            std::shared_ptr<double> point=std::make_shared<double>(aggFac);
+            return point;
+        }
+    }
+    return nullptr;
+
+}
+std::shared_ptr<Vector2> DefencePositionCoach::pickNewPosition(PossiblePass pass,
+        const world::WorldData &world) {
+    std::shared_ptr<Vector2> point = nullptr;
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: please hit ctrl+alt+l

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp [hunk 5]
```diff
@@ -322,6 +327,65 @@ Vector2 DefencePositionCoach::findPositionForBlockBall(const Line &blockLine) {
...
+    double passLen=pass.distance();
+    double segments=30.0;
+    for (int j = 0; j <=segments/2.0; ++ j) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: `segments/2.0` seems very fishy. what are you doing?

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp [hunk 6]
```diff
@@ -322,6 +327,65 @@ Vector2 DefencePositionCoach::findPositionForBlockBall(const Line &blockLine) {
...
+bool DefencePositionCoach::validNewPosition(const Vector2 &position, const world::WorldData &world) {
+    double collisionRadius = 0.15;// a little smaller than 2 robot radii so that we can make solid walls still
+    for (const auto &robot: world.us) {
+        if ((robot.pos - position).length() < collisionRadius) {
+            return false;
+        }
+    }
+    return true;
+}
+std::shared_ptr<double> DefencePositionCoach::pickNewPosition(const Line& line,
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: for the whole file: add more comments. it is quite unclear what you do.

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp [hunk 7]
```diff
@@ -322,6 +327,65 @@ Vector2 DefencePositionCoach::findPositionForBlockBall(const Line &blockLine) {
...
+        const world::WorldData &world) {
+    double lineLen=(line.second-line.first).length();
+    double minSkipDist=0.05;
+    double skipDist=max(minSkipDist,lineLen/30.0);
+    // try positions on the line by skipping forward skipDist every time. If one works return
+    for (int aggFac = 0; aggFac <1/skipDist; ++ aggFac) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: `aggFac`?

Comment (mrlukasbos) [reply roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: this is a whole bunch of magic here

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp [hunk 8]
```diff
@@ -142,12 +143,16 @@ std::vector<DefencePositionCoach::DefenderBot> DefencePositionCoach::decidePosit
...
         // if that doesn't work try to block the pass itself directly
         if (auto blockLine = blockToGoalLine(passes[0], simulatedWorld)) {
-            defender = createBlockToGoal(passes[0], 0.0, *blockLine);
+            if (auto aggressionFactor = pickNewPosition(*blockLine, simulatedWorld)) {
+                defender = createBlockToGoal(passes[0], *aggressionFactor, *blockLine);
+            }
         }
         else if (auto blockPos = blockOnDefenseAreaLine(passes[0], simulatedWorld)) {
-            defender = createBlockOnLine(passes[0], *blockPos);
+            if (validNewPosition(*blockPos, simulatedWorld)) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: Please do not assign variables inside conditionals. That is so ugly....

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp (hunk 1 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: made a const double in the header.

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp (hunk 3 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: fixed it. Think I was tired here, this seems stupid :P

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp (hunk 5 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: added a comment now to explain a bit; I'm searching positions from the middle of the line outwards.

File: roboteam_ai/src/coach/defence/DefencePositionCoach.cpp (hunk 7 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: read the comment, it should be quite clear; we pick a position on the line by trying out positions from the start and checking every time whether or not they are valid. aggFac=aggressionFactor (renamed this)

Comment (rolfvdhulst) [reply roboteam_ai/src/coach/defence/DefencePositionCoach.cpp]: I added a small comment

**Pull request merged**